cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,426 @@
+   * sorted in ascending order.
+   * @param null_precedence Optional, device array the same length as a row and indicates how null
+   * values compare to all other for every column. If it is nullptr, then null precedence would be
+   * `null_order::BEFORE` for all columns.
+   * @param stream The stream to launch kernels and h->d copies on while preprocessing.
+   */
+  preprocessed_table(table_view const& table,
+                     host_span<order const> column_order,
+                     host_span<null_order const> null_precedence,
+                     rmm::cuda_stream_view stream);
+
+  // TODO: Should we add a static create method that returns a shared_ptr?
+
+ private:
+  friend class self_comparator;
+
+  /**
+   * @brief Implicit conversion operator to a `table_device_view` of the preprocessed table.
+   *
+   * @return table_device_view
+   */
+  operator table_device_view() { return **d_t; }
+
+  /**
+   * @brief Get a device array containing the desired order of each column in the preprocessed table
+   *
+   * @return std::optional<device_span<order const>> Device array containing respective column
+   * orders. If no explicit column orders were specified during the creation of this object then
+   * this will be `nullopt`.
+   */
+  [[nodiscard]] std::optional<device_span<order const>> column_order() const
+  {
+    return d_column_order.size() ? std::optional<device_span<order const>>(d_column_order)
+                                 : std::nullopt;
+  }
+
+  /**
+   * @brief Get a device array containing the desired null precedence of each column in the
+   * preprocessed table
+   *
+   * @return std::optional<device_span<null_order const>> Device array containing respective column
+   * null precedence. If no explicit column null precedences were specified during the creation of
+   * this object then this will be `nullopt`.
+   */
+  [[nodiscard]] std::optional<device_span<null_order const>> null_precedence() const
+  {
+    return d_null_precedence.size()
+             ? std::optional<device_span<null_order const>>(d_null_precedence)
+             : std::nullopt;
+  }
+
+  /**
+   * @brief Get a device array containing the depth of each column in the preprocessed table
+   *
+   * @see struct_linearize()
+   *
+   * @return std::optional<device_span<int const>> Device array containing respective column depths.
+   * If there are no nested columns in the table then this will be `nullopt`.
+   */
+  [[nodiscard]] std::optional<device_span<int const>> depths() const
+  {
+    return d_depths.size() ? std::optional<device_span<int const>>(d_depths) : std::nullopt;
+  }
+
+  /**
+   * @brief Whether the table has any nullable column
+   *
+   */
+  [[nodiscard]] bool has_nulls() const { return _has_nulls; }
+
+ private:
+  using table_device_view_owner =
+    std::invoke_result_t<decltype(table_device_view::create), table_view, rmm::cuda_stream_view>;
```

Comment:
Why do we need the extra level of indirection here (pointer to pointer)?

Reply:
The unique pointer returned by `table_device_view::create` is not default constructible. And the way we construct the stored `table_device_view` is after calling `struct_linearize`. So it's not straightforward to construct it in the initializer list either.

Reply:
Maybe make the constructor private (and delete the default constructor) and instead expose a factory `preprocessed_table::create` that calls `struct_initialize` before calling the private constructor? Alternatively, try what Jake suggested below? Just looking for ways to avoid the extra indirection since it is a WAR rather than a necessary feature of the object.

Reply:
I went with your method because it also checks off my [TODO](https://github.com/rapidsai/cudf/blob/096593f68adc460060e4379db54cbca2650a2921/cpp/include/cudf/table/experimental/row_operators.cuh#L282).